internal/xcuitest/config.go

```diff
@@ -0,0 +1,129 @@
+}
+
+// TestOption represents the xcuitest test filter options configuration.
+type TestOption struct {
+	Class []string `yaml:"class,omitempty" json:"class"`
+}
+
+// Suite represents the xcuitest test suite configuration.
+type Suite struct {
+	Name        string          `yaml:"name,omitempty" json:"name"`
+	Devices     []config.Device `yaml:"devices,omitempty" json:"devices"`
+	TestOptions TestOption      `yaml:"testOptions,omitempty" json:"testOptions"`
+}
+
+// FromFile creates a new xcuitest Project based on the filepath cfgPath.
+func FromFile(cfgPath string) (Project, error) {
+	var p Project
+
+	f, err := os.Open(cfgPath)
+	if err != nil {
+		return Project{}, fmt.Errorf("failed to locate project config: %v", err)
+	}
+	defer f.Close()
+
+	if err := yaml.NewDecoder(f).Decode(&p); err != nil {
+		return Project{}, fmt.Errorf("failed to parse project config: %v", err)
+	}
+	p.ConfigFilePath = cfgPath
+
+	if p.Sauce.Concurrency < 1 {
+		// Default concurrency is 2
+		p.Sauce.Concurrency = 2
+	}
+
+	return p, nil
+}
+
+// Validate validates basic configuration of the project and returns an error if any of the settings contain illegal
+// values. This is not an exhaustive operation and further validation should be performed both in the client and/or
+// server side depending on the workflow that is executed.
+func Validate(p Project) error {
+	if p.Xcuitest.App == "" {
+		return errors.New("missing path to app .ipa")
+	}
+	if !strings.HasSuffix(p.Xcuitest.App, ".ipa") {
+		return fmt.Errorf("invaild application file: %s, make sure extension is .ipa", p.Xcuitest.App)
+	}
+
+	if p.Xcuitest.TestApp == "" {
+		return errors.New("missing path to test app .ipa")
+	}
+	if !strings.HasSuffix(p.Xcuitest.TestApp, ".ipa") {
+		return fmt.Errorf("invaild application test file: %s, make sure extension is .ipa", p.Xcuitest.TestApp)
+	}
+
+	if len(p.Suites) == 0 {
+		return errors.New("no suites defined")
+	}
+
+	for _, suite := range p.Suites {
+		if len(suite.Devices) == 0 {
+			return fmt.Errorf("missing devices configuration for suite: %s", suite.Name)
+		}
+		for didx, device := range suite.Devices {
+			if device.ID == "" && device.Name == "" {
+				return fmt.Errorf("missing device name or id for suite: %s. Devices index: %d", suite.Name, didx)
+			}
+
+			if device.Options.DeviceType != "" && !isSupportedDeviceType(device.Options.DeviceType) {
+				return fmt.Errorf("deviceType: %s is unsupported for suited: %s. Devices index: %d. Supported device types: %s",
+					device.Options.DeviceType, suite.Name, didx, strings.Join(supportedDeviceTypes, ","))
+			}
+		}
+	}
+
+	return nil
+}
+
+// ApplyDefaultValues applies default values.
+func ApplyDefaultValues(p *Project) {
+	for _, suite := range p.Suites {
+		for _, device := range suite.Devices {
+			device.PlatformName = "iOS"
+
+			// device type only supports uppercase values
+			if device.Options.DeviceType != "" {
+				device.Options.DeviceType = strings.ToUpper(device.Options.DeviceType)
+			}
```

Comment:
This part bother me because it is not applying a default value per-se.
May be this should be placed somewhere else or rename the function.

Reply:
I renamed it to `SetDeviceDefaultValues` as it's related to the device only.